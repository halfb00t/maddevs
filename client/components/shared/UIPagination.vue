<template>
  <div class="pagination">
    <ul
      class="pagination__list"
      :class="{'pagination__list--few-items': totalPages <= 3}"
    >
      <li class="pagination__list-item">
        <button
          class="pagination__list-arrow-btn"
          :disabled="isInFirstPage"
          :title="isInFirstPage && 'No pages'"
          aria-label="Go to previous page"
          @click="onClickPreviousPage"
        >
          &lt;--- Prev
        </button>
      </li>
      <li
        v-for="page in pages"
        :key="page.name"
      >
        <button
          type="button"
          :disabled="page.isDisabled"
          :title="page.isDisabled && 'This is current page'"
          class="pagination__list-pages-btn"
          :class="{ 'pagination__list-pages-btn--active': isPageActive(page.name) }"
          :aria-label="`Go to page number ${page.name}`"
          @click="onClickPage(page.name)"
        >
          {{ page.name }}
        </button>
      </li>
      <li v-if="totalPages > currentPage + 1">
        <span
          class="pagination__list-pages-btn"
          style="cursor: default;"
        >
          ...
        </span>
      </li>
      <li v-if="totalPages > currentPage + 1">
        <button
          type="button"
          :disabled="currentPage === totalPages"
          class="pagination__list-pages-btn"
          :class="{ 'pagination__list-pages-btn--active': isPageActive(totalPages) }"
          :aria-label="`Go to page number ${totalPages}`"
          @click="onClickPage(totalPages)"
        >
          {{ totalPages }}
        </button>
      </li>
      <li class="pagination__list-item">
        <button
          class="pagination__list-arrow-btn"
          :disabled="isInLastPage"
          :title="isInLastPage && 'No pages'"
          aria-label="Go to next page"
          @click="onClickNextPage"
        >
          Next ---&gt;
        </button>
      </li>
    </ul>
  </div>
</template>
<script>
export default {
  name: 'UIPagination',

  props: {
    maxVisibleButtons: {
      type: Number,
      default: 3,
    },

    totalPages: {
      type: Number,
      required: true,
    },

    perPage: {
      type: Number,
      default: 6,
    },

    currentPage: {
      type: Number,
      required: true,
    },
  },

  computed: {
    startPage() {
      if (this.currentPage === 1) return 1
      if (this.currentPage === this.totalPages && this.totalPages >= this.maxVisibleButtons) {
        return this.totalPages - this.maxVisibleButtons + 1
      }
      return this.currentPage - 1
    },

    pages() {
      const range = []

      if (this.totalPages >= this.maxVisibleButtons) {
        for (let i = this.startPage;
          i <= Math.min(this.startPage + this.maxVisibleButtons - 1, this.totalPages);
          i += 1) {
          range.push({
            name: i,
            isDisabled: i === this.currentPage,
          })
        }
      } else {
        for (let i = this.startPage; i <= this.totalPages && i > 0; i += 1) {
          range.push({
            name: i,
            isDisabled: i === this.currentPage,
          })
        }
      }
      return range
    },

    isInFirstPage() {
      return this.currentPage === 1
    },

    isInLastPage() {
      return this.currentPage === this.totalPages
    },
  },

  methods: {
    onClickPreviousPage() {
      this.$emit('pageChanged', this.currentPage - 1)
    },

    onClickPage(page) {
      this.$emit('pageChanged', page)
    },

    onClickNextPage() {
      this.$emit('pageChanged', this.currentPage + 1)
    },

    isPageActive(page) {
      return this.currentPage === page
    },
  },
}
</script>
<style lang="scss" scoped>
%btn {
  background-color: transparent;
  border: none;
  @include font('Inter', 16px, 400);
  line-height: 27px;
  cursor: pointer;
}

.pagination {
  max-width: 311px;
  margin: 0 auto;
  &__list {
    display: flex;
    justify-content: space-between;
    align-items: center;
    &--few-items {
      justify-content: space-evenly;
    }
  }
  &__list-arrow-btn {
    @extend %btn;
    color: $text-color--red;
    &:disabled {
      cursor: not-allowed;
      color: $text-color--grey-20-percent;
    }
  }

  &__list-pages-btn {
    @extend %btn;
    color: $text-color--grey-20-percent;
    outline: 5px solid transparent !important;
    &--active {
      cursor: not-allowed;
      color: $text-color--red;
    }
  }
}
</style>
