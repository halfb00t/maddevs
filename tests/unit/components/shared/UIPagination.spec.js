import { fireEvent, render, screen } from '@testing-library/vue'
import { shallowMount } from '@vue/test-utils'
import UIPagination from '@/components/shared/UIPagination'

const props = {
  maxVisibleButtons: 3,
  totalPages: 4,
  perPage: 6,
  currentPage: 1,
}

describe('UIPagination component', () => {
  let wrapper

  afterEach(() => {
    wrapper = null
  })

  it('should correctly render component', () => {
    const { container } = render(UIPagination, {
      props,
    })

    expect(container).toMatchSnapshot()
  })

  it('should correctly return computed method', () => {
    props.currentPage = 4
    wrapper = shallowMount(UIPagination, {
      propsData: props,
    })

    expect(wrapper.vm.startPage).toBe(2)
  })

  it('should correctly return startPage computed', () => {
    props.currentPage = 5
    wrapper = shallowMount(UIPagination, {
      propsData: props,
    })

    expect(wrapper.vm.startPage).toBe(4)
  })

  it('should correctly render with totalPages < 3', () => {
    props.currentPage = 1
    props.totalPages = 2

    wrapper = shallowMount(UIPagination, {
      propsData: props,
    })

    expect(wrapper.vm.pages).toHaveLength(props.totalPages)
  })

  it('should correctly click to prev page', () => {
    const $emit = jest.fn()
    props.currentPage = 2

    wrapper = shallowMount(UIPagination, {
      propsData: props,
      mocks: {
        $emit,
      },
    })

    const buttons = wrapper.find('.pagination__list-arrow-btn')
    buttons.trigger('click')
    expect($emit).toHaveBeenCalledWith('pageChanged', 1)
  })

  it('should correctly click to next page', async () => {
    const $emit = jest.fn()
    props.currentPage = 1
    render(UIPagination, {
      propsData: props,
      mocks: {
        $emit,
      },
    })

    const nextBtn = screen.getByTestId('text-next-btn')
    await fireEvent.click(nextBtn)
    expect($emit).toHaveBeenCalledWith('pageChanged', 2)
  })

  it('should correctly click to page', () => {
    const $emit = jest.fn()
    wrapper = shallowMount(UIPagination, {
      propsData: props,
      mocks: {
        $emit,
      },
    })

    const buttons = wrapper.findAll('.pagination__list-pages-btn')
    buttons.at(1).trigger('click')
    expect($emit).toHaveBeenCalledWith('pageChanged', 2)
  })
})
