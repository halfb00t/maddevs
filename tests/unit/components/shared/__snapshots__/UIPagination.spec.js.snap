// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIPagination component should correctly render component 1`] = `
<div>
  <div
    class="pagination"
  >
    <ul
      class="pagination__list"
    >
      <li
        class="pagination__list-item"
      >
        <button
          aria-label="Go to previous page"
          class="pagination__list-arrow-btn"
          disabled="disabled"
          title="No pages"
        >
          
        &lt;--- Prev
      
        </button>
      </li>
       
      <li>
        <button
          aria-label="Go to page number 1"
          class="pagination__list-pages-btn pagination__list-pages-btn--active"
          disabled="disabled"
          title="This is current page"
          type="button"
        >
          
        1
      
        </button>
      </li>
      <li>
        <button
          aria-label="Go to page number 2"
          class="pagination__list-pages-btn"
          type="button"
        >
          
        2
      
        </button>
      </li>
      <li>
        <button
          aria-label="Go to page number 3"
          class="pagination__list-pages-btn"
          type="button"
        >
          
        3
      
        </button>
      </li>
       
      <li>
        <span
          class="pagination__list-pages-btn"
          style="cursor: default;"
        >
          
        ...
      
        </span>
      </li>
       
      <li>
        <button
          aria-label="Go to page number 4"
          class="pagination__list-pages-btn"
          type="button"
        >
          
        4
      
        </button>
      </li>
       
      <li
        class="pagination__list-item"
      >
        <button
          aria-label="Go to next page"
          class="pagination__list-arrow-btn"
          data-testid="text-next-btn"
        >
          
        Next ---&gt;
      
        </button>
      </li>
    </ul>
  </div>
</div>
`;
